packages/SwingSet/src/kernel/vat-loader/manager-subprocess-xsnap.js

```diff
@@ -215,19 +227,60 @@ export function makeXsSubprocessFactory({
     mk.setDeliverToWorker(deliverToWorker);
 
     function shutdown() {
+      handleCommandKit?.revoke();
+      handleCommandKit = undefined;
       return worker.close().then(_ => undefined);
     }
     /**
      * @param {number} snapPos
      * @param {SnapStore} snapStore
+     * @param {boolean} [keepWorker]
      * @returns {Promise<SnapshotResult>}
      */
-    function makeSnapshot(snapPos, snapStore) {
-      return snapStore.saveSnapshot(
-        vatID,
+    async function makeSnapshot(snapPos, snapStore, keepWorker) {
+      const snapshotDescription = `${vatID}-${snapPos}`;
+      const snapshotStream = worker.makeSnapshotStream(snapshotDescription);
+
+      if (keepWorker) {
+        return snapStore.saveSnapshot(vatID, snapPos, snapshotStream);
+      }
+
+      /** @type {AsyncGenerator<Uint8Array, void, void>[]} */
+      const [loadStream, saveStream] = synchronizedTee(snapshotStream, 2);
```

Comment:
I might name these `restartWorkerStream` and `dbSaveStream` (or `saveToDBStream`), the "load vs save" directions are a bit ambiguous because they don't tell you it's the replacement worker doing the loading.